Modularize kernel modules manager into its own cell

pippolo84 opened the pull request:
Modularize the kernel modules manager into its own cell.

Instead of exporting an Init() function, the manager stores the list of loaded modules once at startup.

The tests are updated accordingly and since we are at it, migrate them from checkmate (the temporary wrapper for gopkg.in/check.v1) to the standard Go testing framework.

This is a first step needed to modularize the iptables manager.

pippolo84 committed:
iptables: Remove unused KernelHasNetfilter func

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

pippolo84 committed:
pkg/modules: Modularize kernel modules manager

Modularize the kernel modules manager into its own cell.

Instead of an exported Init() function, the manager stores the list of
loaded modules once at startup.

The tests are updated accordingly and since we are at it, migrate them
from checkmate (the temporary wrapper for gopkg.in/check.v1) to the
standard Go testing framework.

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

YutaroHayakawa reviewed (approved):
LGTM 👍

joamaki reviewed (changes requested):
Oops, reviewed this in another PR that depended on this. See comments there: https://github.com/cilium/cilium/pull/28746

pippolo84 committed:
linux/modules: Move package under datapath/linux

Kernel modules manager is meant to work with Linux kernel modules, so
it belongs to the datapath/linux folder.

CODEOWNERS file is updated accordingly.

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

joamaki commented on pkg/datapath/cells.go:
```diff
@@ -52,6 +53,10 @@ var Cell = cell.Module(
 	monitorAgent.Cell,
 
+	// The modules manager to search and load kernel modules.
+	// Must appear before the datapath so that its start hook is registered before the iptables one.
```
This comment about ordering is not true. Since `newDatapath` depends on the modules manager it will be started before iptables. Also since `modules.Cell` does not have an invoke, it doesn't matter where it is here since the order only matters for invokes.

pippolo84 replied:
Got confused with `Invoke` and `OnStart`, I'm gonna fix this :+1:

joamaki replied:
Thanks! :smiley:

joamaki commented on pkg/datapath/linux/modules/modules.go:
```diff
@@ -50,3 +45,11 @@ func (m *ModulesManager) FindOrLoadModules(expectedNames ...string) error {
+func newManager(lc hive.Lifecycle) *Manager {
+	mm := Manager{}
+
+	lc.Append(hive.Hook{OnStart: mm.Start})
```
Hmm, I'm not sure what this does. `mm` is `Manager` and not `*Manager`, so there might be a copy happening here. Please do `m := &Manager{}` instead. And please call it `m` instead of `mm` to match method receivers.

pippolo84 replied:
Good catch.

joamaki commented on pkg/datapath/linux/modules/modules_test.go:
```diff
@@ -0,0 +1,23 @@
+	"github.com/cilium/cilium/pkg/hive"
+)
+
+func TestInit(t *testing.T) {
```
This test doesn't seem to do anything useful. `Cell` only has a single `Provide` for an object that is never used by the test, thus it's never constructed. I would just delete this.
